File: 2015/13_KnightsoftheDinnerTable/seating.js
/* eslint-disable linebreak-style */
// ======================================================================
// Knights of the Dinner Table
//   Advent of Code 2015 Day 13 -- Eric Wastl -- https://adventofcode.com
//
// JavaScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           s e a t i n g . j s
//
// A solver for the Advent of Code 2015 Day 13 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------

// ======================================================================
//                                                                Seating
// ======================================================================

class Seating {
  // Object for Knights of the Dinner Table

  constructor(options) {
    // Create a Seating object

    // 1. Set the initial values
    this.text = options.text === undefined ? null : options.text;
    this.part2 = options.part2 === undefined ? false : options.part2;
    this.attendees = {};
    this.number = 0;

    // 2. Process text (if any)
    if (this.text !== null) {
      this.processText(this.text);
    }

    // 3. For part2, the host is included
    if (this.part2) {
      this.number += 1;
    }
  }

  processText(text) {
    // 1. Start with nothing
    this.attendees = {};
    this.number = 0;
    const pattern = /([A-Z][A-Za-z]+) would (gain|lose) ([0-9]+) happiness units by sitting next to ([A-Z][A-Za-z]+)\./;

    // 2. Loop for each line of the input
    text.forEach((line) => {
      // 3. Decompose input line into cities and distance
      const match = line.match(pattern);
      if (match) {
        const personOne = match[1];
        const gainLoss = match[2];
        let amount = parseInt(match[3], 10);
        const personTwo = match[4];

        // 4. Gain or loss
        if (gainLoss === 'lose') {
          amount = -amount;
        }

        // 5. Add the attendee
        this.addAttendee(personOne, amount, personTwo);
      } else {
        // eslint-disable-next-line no-console
        console.log('Unable to parse input', line);
      }
    });
  }

  addAttendee(personOne, amount, personTwo) {
    if (!(personOne in this.attendees)) {
      this.attendees[personOne] = {};
      this.number += 1;
    }
    this.attendees[personOne][personTwo] = amount;
  }

  nextTo(personOne, personTwo) {
    let happiness = 0;
    if (personOne in this.attendees && personTwo in this.attendees[personOne]) {
      happiness += this.attendees[personOne][personTwo];
    }
    if (personTwo in this.attendees && personOne in this.attendees[personTwo]) {
      happiness += this.attendees[personTwo][personOne];
    }
    return happiness;
  }

  totalHappiness(table) {
    if (table === null || table.length !== this.number) {
      return null;
    }
    let happiness = this.nextTo(table[0], table[this.number - 1]);
    for (let i = 0; i < this.number - 1; i += 1) {
      happiness += this.nextTo(table[i], table[i + 1]);
    }
    return happiness;
  }

  maximizeHappiness() {
    // 1. Start with nothing
    let bestTable = null;
    let bestHappiness = null;
    // 2. Get the names of the attendees and vary all but the host
    const keys = Object.keys(this.attendees);
    let host = keys[0];
    let others = keys.slice(1);
    if (this.part2) {
      host = 'yourself';
      others = keys;
    }
    // 3. Get variations of the others
    const variations = Seating.permutator(others);
    // 4. Loop for all of the variations
    for (let i = 0; i < variations.length; i += 1) {
      // 5. Determine the happiness from this table arrangement
      const table = [host].concat(variations[i]);
      const happiness = this.totalHappiness(table);
      // 6. If the first or better than all the previous save it
      if (bestHappiness === null || happiness > bestHappiness) {
        bestHappiness = happiness;
        bestTable = table;
      }
    }
    // 7. Return the most happy of arrangements
    return bestTable;
  }

  // Derived from https://stackoverflow.com/questions/9960908/permutations-in-javascript
  static permutator(inputArr) {
    const result = [];
    const permute = (arr, m = []) => {
      if (arr.length === 0) {
        result.push(m);
      } else {
        for (let i = 0; i < arr.length; i += 1) {
          const curr = arr.slice();
          const next = curr.splice(i, 1);
          permute(curr.slice(), m.concat(next));
        }
      }
    };
    permute(inputArr);
    return result;
  }

  partOne(options) {
    // Returns the solution for part one

    // 0. Function arguments
    // eslint-disable-next-line no-unused-vars
    const verbose = options.verbose === undefined ? false : options.verbose;
    // eslint-disable-next-line no-unused-vars
    const limit = options.limit === undefined ? 0 : options.limit;

    // 1. Return the solution for part one
    return this.totalHappiness(this.maximizeHappiness());
  }

  partTwo(options) {
    // Returns the solution for part two

    // 0. Function arguments
    // eslint-disable-next-line no-unused-vars
    const verbose = options.verbose === undefined ? false : options.verbose;
    // eslint-disable-next-line no-unused-vars
    const limit = options.limit === undefined ? 0 : options.limit;

    // 1. Return the solution for part two
    return this.totalHappiness(this.maximizeHappiness());
  }
}

module.exports.Seating = Seating;
// ======================================================================
// end                       s e a t i n g . j s                      end
// ======================================================================

File: 2015/13_KnightsoftheDinnerTable/seating.py
# ======================================================================
# Knights of the Dinner Table
#   Advent of Code 2015 Day 13 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         s e a t i n g . p y
# ======================================================================
"A solver for the Advent of Code 2015 Day 13 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import re
from itertools import permutations

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
RE_SEATING = re.compile("([A-Z][a-z]+) would (gain|lose) ([0-9]+) "
                        + "happiness units by sitting next to ([A-Z][a-z]+).")

# ======================================================================
#                                                                Seating
# ======================================================================


class Seating(object):   # pylint: disable=R0902, R0205
    "Object for Knights of the Dinner Table"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.people = set()
        self.preferences = {}

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            for line in text:
                self._process_line(line)

    def _process_line(self, line):
        "Parse and assign values from the line"

        # 1. Parse the line
        match = RE_SEATING.match(line)
        if not match:
            print("Unable to parse line:", line)
            return

        # 2. Get the interesting bits
        who, gain, amount, whom = match.groups()
        amount = int(amount)
        if gain == "lose":
            amount = -amount

        # 3. Save the information
        self.people.add(who)
        self.people.add(whom)
        self.preferences[(who, whom)] = amount

    def optimum(self, verbose=False):
        "Seat the people optimumly"

        # 1. Start with very low expectations
        result = -9999999
        best = []
        if len(self.people) < 2:
            return result

        # 2. Loop for all of the possible seatings
        for seating_chart in permutations(self.people):

            # 3. Determine the happiness score for this seating chart
            score = self.total_happiness(seating_chart)

            # 4. If this one is better, save it
            if score > result:
                result = score
                best = seating_chart

        # 5. Return the result
        if verbose:
            print("Best", best, result)
        return result

    def total_happiness(self, seating_chart):
        "Return the total happiness of this seating chart"

        # 1. Start with nothing
        result = 0

        # 2. Loop for all the pairs of people
        for indx in range(len(self.people) - 1):

            # 3. Score this pair of people
            result += self.preferences[(seating_chart[indx], seating_chart[indx + 1])]
            result += self.preferences[(seating_chart[indx + 1], seating_chart[indx])]

        # 4. Don't forget that the table is circular
        result += self.preferences[(seating_chart[0], seating_chart[-1])]
        result += self.preferences[(seating_chart[-1], seating_chart[0])]

        # 5. Return the score for this seating chart
        return result

    def seat_yourself(self):
        "Add yourself into the seating situation"

        # 1. Loop for all of the people
        for person in self.people:

            # 2. You are neutral about them and they are neutral about you
            self.preferences[("Me", person)] = 0
            self.preferences[(person, "Me")] = 0

        # 3. You are a person
        self.people.add("Me")

    def part_one(self, verbose=False, limit=0):
        "Returns the solution for part one"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part one
        return self.optimum(verbose=verbose)

    def part_two(self, verbose=False, limit=0):
        "Returns the solution for part two"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part two
        self.seat_yourself()
        return self.optimum(verbose=verbose)


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                       s e a t i n g . p y                      end
# ======================================================================
